cli/replicaset/cmd/upgrade.go

```diff
@@ -0,0 +1,279 @@
+	for _, instance := range replicaset.Instances {
+		run := instance.InstanceCtx
+		fullInstanceName := running.GetAppInstanceName(run)
+		conn, err := getInstanceConnector(instance)
+
+		if err != nil {
+			return err
+		}
+
+		var isRW bool
+		if instance.Mode.String() != "unknown" {
+			isRW = instance.Mode.String() == "rw"
+		} else {
+			res, err := conn.Eval(
+				"return (type(box.cfg) == 'function') or box.info.ro",
+				[]any{}, connector.RequestOpts{})
+			if err != nil {
+				return fmt.Errorf("[%s]: %w", fullInstanceName, err)
+			}
+			isRW = !res[0].(bool)
+		}
+
+		if isRW && *master != nil {
+			return fmt.Errorf("%s and %s are both masters",
+				running.GetAppInstanceName((*master).run), fullInstanceName)
+		} else if isRW {
+			*master = &InstanceMeta{run, conn}
+		} else {
+			*replicas = append(*replicas, InstanceMeta{run, conn})
+		}
+	}
+	return nil
+}
+
+func WaitLSN(conn connector.Connector, masterIID uint32, masterLSN uint64) error {
+	var lastError error
+	query := fmt.Sprintf("return box.info.vclock[%d]", masterIID)
+
+	deadline := time.Now().Add(time.Duration(LsnTimeout) * time.Second)
+	for {
+		res, err := conn.Eval(query, []any{}, connector.RequestOpts{})
+		if err != nil {
+			lastError = fmt.Errorf("failed to evaluate LSN query: %w", err)
+		} else if len(res) == 0 {
+			lastError = errors.New("empty result from LSN query")
+		} else {
+			var lsn uint64
+			if err := mapstructure.Decode(res[0], &lsn); err != nil {
+				lastError = fmt.Errorf("failed to decode LSN: %w", err)
+			} else if lsn >= masterLSN {
+				return nil
+			} else {
+				lastError = fmt.Errorf("current LSN %d is behind required "+
+					"master LSN %d", lsn, masterLSN)
+			}
+		}
+
+		if time.Now().After(deadline) {
+			break
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+
+	return lastError
+}
+
+func upgradeMaster(master *InstanceMeta) (SyncInfo, error) {
+	var upgradeInfo SyncInfo
+	fullMasterName := running.GetAppInstanceName(master.run)
+	res, err := master.conn.Eval(upgradeMasterLua, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return upgradeInfo, fmt.Errorf(
+			"failed to execute upgrade script on master instance - %s: %w",
+			fullMasterName, err)
+	}
+
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return upgradeInfo, fmt.Errorf(
+			"failed to decode response from master instance - %s: %w",
+			fullMasterName, err)
+	}
+
+	if upgradeInfo.Err != nil {
+		return upgradeInfo, fmt.Errorf(
+			"master instance upgrade failed - %s: %w",
+			fullMasterName, err)
+	}
+	return upgradeInfo, nil
+}
+
+func Snapshot(instance *InstanceMeta) error {
+	res, err := instance.conn.Eval("return box.snapshot()", []any{},
+		connector.RequestOpts{})
+	if err != nil {
+		return fmt.Errorf("failed to execute snapshot on replica: %w", err)
+	}
+	if len(res) == 0 {
+		return fmt.Errorf("snapshot command on %s returned an empty result, "+
+			"'ok' - expected", running.GetAppInstanceName(instance.run))
```

Comment:
If 'ok' expected than we need to check the value.

Reply:
Done.